awx/main/migrations/_galaxy.py

```diff
@@ -0,0 +1,85 @@
+# Generated by Django 2.2.11 on 2020-08-04 15:19
+
+import logging
+
+from awx.main.utils.encryption import encrypt_field, decrypt_field
+
+from django.utils.timezone import now
+
+from awx.main.models import CredentialType as ModernCredentialType
+from awx.main.utils.common import set_current_apps
+
+logger = logging.getLogger('awx.main.migrations')
+
+
+def migrate_galaxy_settings(apps, schema_editor):
+    set_current_apps(apps)
+    ModernCredentialType.setup_tower_managed_defaults()
+    Organization = apps.get_model('main', 'Organization')
+    CredentialType = apps.get_model('main', 'CredentialType')
+    Credential = apps.get_model('main', 'Credential')
+    Setting = apps.get_model('conf', 'Setting')
+
+    galaxy_type = CredentialType.objects.get(kind='galaxy')
+    private_galaxy_url = Setting.objects.filter(key='PRIMARY_GALAXY_URL').first()
+
+    # by default, prior versions of AWX/Tower automatically pulled content
+    # from galaxy.ansible.com
+    public_galaxy_enabled = True
+    public_galaxy_setting = Setting.objects.filter(key='PUBLIC_GALAXY_ENABLED').first()
+    if public_galaxy_setting and public_galaxy_setting.value is False:
+        # ...UNLESS this behavior was explicitly disabled via this setting
+        public_galaxy_enabled = False
+
+    for org in Organization.objects.all():
+        if private_galaxy_url and private_galaxy_url.value:
+            # If a setting exists for a private Galaxy URL, make a credential for it
+            username = Setting.objects.filter(key='PRIMARY_GALAXY_USERNAME').first()
+            password = Setting.objects.filter(key='PRIMARY_GALAXY_PASSWORD').first()
+            if (username and username.value) or (password and password.value):
+                logger.error(
+                    f'Specifying HTTP basic auth for the Ansible Galaxy API '
+                    f'({private_galaxy_url.value}) is no longer supported. '
+                    'Please provide an API token instead after your upgrade '
+                    'has completed',
+                )
+            inputs = {
+                'url': private_galaxy_url.value
+            }
+            token = Setting.objects.filter(key='PRIMARY_GALAXY_TOKEN').first()
+            if token and token.value:
+                inputs['token'] = decrypt_field(token, 'value')
+            auth_url = Setting.objects.filter(key='PRIMARY_GALAXY_AUTH_URL').first()
+            if auth_url and auth_url.value:
+                inputs['auth_url'] = auth_url.value
+            name = f'Private Galaxy ({private_galaxy_url.value})'
+            if 'cloud.redhat.com' in inputs['url']:
+                name = f'Ansible Automation Hub ({private_galaxy_url.value})'
+            cred = Credential(
+                created=now(),
+                modified=now(),
+                name=name,
+                organization=org,
+                credential_type=galaxy_type,
+                inputs=inputs
+            )
+            cred.save()
```

Comment:
I am going to have a lot of RBAC questions that stem from this. Before we even get into any feature-specific discussions, does this correctly populate the credential's role fields and their parents/ancestors? I don't trust that it does in migrations.

You made the organization of the credential the same as the organization it gets associated. Do you intend to enforce this for new credentials?

Reply:
Good questions - this is an area where our usage of credentials here is a _little_ weird.  It doesn't really make sense to create one of these credentials outside of the context of its association with at least one Organization (i.e.., there's no reason for somebody lower than Org Admin to make one of these, because they can't actually assign it to an Org).

Reply:
>  does this correctly populate the credential's role fields and their parents/ancestors?

It _seems_ like it does?

```In [8]: Credential.objects.first().name
Out[8]: 'Ansible Galaxy'

In [9]: Credential.objects.first().inputs
Out[9]: {'url': 'https://galaxy.ansible.com/'}

In [10]: Credential.objects.first().admin_role.ancestors.all()
Out[10]: <QuerySet [<Role: Admin-36>, <Role: Admin-2>, <Role: Credential Admin-6>, <Role: System Administrator-1>]>````